### PR TITLE
volume snapshot delete and restore support

### LINKED ISSUE #10259: ws-manager: add support for deleting snapshotVolumes (to clean up old backups) on workspace delete

### PR DESCRIPTION
**Pull request by sagor999**
## Description

This PR adds two features:
1. New workspace volume snapshot is created. We need to delete old volume snapshot for that workspace now.
2. Open workspace in the cluster where volume snapshot doesn't exist yet. We need to restore volume snapshot.
3. Workspace is deleted: it will garbage collect all volume snapshots associated with that workspace.

Fixes #10259 

## How to test

Loom video showing this in action: 
https://www.loom.com/share/51b41802054d4de89902aaa358171624

## Release Notes

```release-note
none
```

### CODE REVIEW

**Comment by roboquat** [comment]
@sagor999: Adding the "do-not-merge/release-note-label-needed" label because no release-note block was detected, please follow our [release note process](https://git.k8s.io/community/contributors/guide/release-notes.md) to remove it.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

File: components/ws-manager-api/core.proto
Comment (jenting) [inline components/ws-manager-api/core.proto:43]: Nit: it'd be good to have a comment describing this function.

File: components/ws-manager/pkg/manager/manager.go
Comment (jenting) [inline components/ws-manager/pkg/manager/manager.go:425]: Nit: We could consider add VolumeSnapshotClassName to VolumeSnapshotContent as well.
```go
if volumeSnapshotClassName != "" {
	volumeSnapshotContent.Spec.VolumeSnapshotClassName = &volumeSnapshotClassName
}
```

File: components/ws-manager/pkg/manager/manager.go
Comment (jenting) [inline components/ws-manager/pkg/manager/manager.go:367]: nil: we could consider remove this line

File: components/ws-manager/pkg/manager/manager.go
Comment (jenting) [inline components/ws-manager/pkg/manager/manager.go:438]: nit: we could consider remove this line

File: components/ws-manager/pkg/manager/manager.go
Comment (jenting) [inline components/ws-manager/pkg/manager/manager.go:446]: Not quite understand that _why_ do we need to support soft delete?
The purpose seems like we will restore the VolumeSnapshotContent as well as VolumeSnapshot from the GCP snapshots, then delete the VolumeSnapshot object. Therefore, the VolumeSnapshotContent remains.

Could you please elaborate more for the purpose?

Comment (sagor999) [reply components/ws-manager/pkg/manager/manager.go:446]: this is for a future PR:
when we need to remove workspace and all associated volume snapshots for it, that could be spread over multiple clusters. 
So I was thinking I could send request to soft delete to all clusters except one that will actually perform hard delete.

File: components/ws-manager/pkg/manager/manager.go
Comment (sagor999) [reply components/ws-manager/pkg/manager/manager.go:425]: Not sure I understood what you meant.

File: components/ws-manager/pkg/manager/manager.go
Comment (sagor999) [reply components/ws-manager/pkg/manager/manager.go:438]: why this one? I like to have a field with func name where the log entry is coming from, as sometimes our log output text can be repeated in multiple places and then it might be difficult at times to figure out where this specific entry is coming from.

Comment (jenting) [reply components/ws-manager/pkg/manager/manager.go:437]: Okay, that makes sense. Let's keep it as it is.

File: components/ws-manager/pkg/manager/manager.go
Comment (jenting) [reply components/ws-manager/pkg/manager/manager.go:424]: Because we add volumeSnapshotClassName to VolumeSnapshot object on lines 397-399.
I'm thinking that we could add volumeSnapshotClassName to VolumeSnapshotContent object as well. (Although if we don't specify the VolumeSnapshotClass, the Kubernetes uses the default one within the cluster).

**Comment by geropl** [comment]
Starting to review now

File: components/ws-manager-bridge/src/bridge.ts
Comment (geropl) [inline components/ws-manager-bridge/src/bridge.ts:384]: Uff. IMo the call _back_ to ws-manager-bridge for cleanup purposes is something we should _not_ handle here. :confused: It's already complicated and untestable as-is, but we should try hard to not make it worse.

Instead, I propose we have add a garbage-collection mechanism to `server`, in the same style that we use for workspaces, for instance. Optionally, this can be triggered by "workspace stopped" events. But has the additional benefit of being more reliable.

@sagor999 WDYT?

File: components/ws-manager/pkg/manager/manager.go
Comment (sagor999) [reply components/ws-manager/pkg/manager/manager.go:445]: @jenting there is an example of soft delete usage now in this PR.

File: components/ws-manager-bridge/src/bridge.ts
Comment (sagor999) [reply components/ws-manager-bridge/src/bridge.ts]: @geropl feedback addressed!

**Comment by sagor999** [comment]
/hold
to ensure I get review from webapp team.

File: components/ws-manager/pkg/manager/manager.go
Comment (jenting) [inline components/ws-manager/pkg/manager/manager.go:470]: Should this one be `if k8serr.IsNotFound(err) {`?

File: components/server/src/workspace/garbage-collector.ts
Comment (geropl) [inline components/server/src/workspace/garbage-collector.ts:136]: Note: This is potentially unbound, and might blow up the whole process as we grow.
The other deletion mechanisms feature a "limit" which I think we could re-use here.

File: components/server/src/workspace/workspace-deletion-service.ts
Comment (geropl) [inline components/server/src/workspace/workspace-deletion-service.ts:122]: > gloud

cloud

File: components/server/src/workspace/workspace-deletion-service.ts
Comment (geropl) [inline components/server/src/workspace/workspace-deletion-service.ts:140]: I expected this method to connect to a specific ws-manager based on a "region" field, either retrieved from `VolumeSnapshot`, `Workspace` or `WorkspaceInstance`. Iterating over all currently connected clients makes it impossible to detect cases where we are no longer connected to a certain ws-manager.

Or is this due to the fact that we might have multiple ws-manager (with different names) in the same GCloud region, and we do not know how to map those? :thinking:

**Comment by geropl** [comment]
@sagor999 Regarding merging of this PR: Once the initial review is through, it'd would be awesome to decouple this PR into smaller chunks. For instance:
 - DB changes
 - ws-manager changes
 - server+bridge changes

File: components/gitpod-db/src/typeorm/migration/1654628106102-VolumeSnapshotAddWSId.ts
Comment (geropl) [inline components/gitpod-db/src/typeorm/migration/1654628106102-VolumeSnapshotAddWSId.ts:16]: > ind_5iziusbz752p4jv3e8erq3wc7n

We usually use readable names ala `ind_workspaceId` :+1:

File: components/gitpod-db/src/typeorm/workspace-db-impl.ts
Comment (geropl) [inline components/gitpod-db/src/typeorm/workspace-db-impl.ts:781]: Checking this query it again it might be necessary to instead of introduce `ind_workspaceId`, make it a combined 2ndary key: `ind_workspaceId_creationTime (workspaceId, creationTime)`. This we guarantee it fits here, but can also be used for workspaceId alone. :+1:

File: components/ws-manager/pkg/manager/manager.go
Comment (sagor999) [reply components/ws-manager/pkg/manager/manager.go:470]: no. we want to report back if we indeed deleted snapshot volume. if it was not found, then nothing was actually deleted.

File: components/server/src/workspace/workspace-deletion-service.ts
Comment (sagor999) [reply components/server/src/workspace/workspace-deletion-service.ts:140]: Not quite. Imagine following scenario:
EU cluster where WS was started from Volume Snapshot X
US cluster where WS was also started from same Volume Snapshot X (from prebuild for example)

Then we want to GC that volume snapshot.
We need to do two things now:
1. Ensure that we delete k8s object representing this snapshot from ALL clusters (to make sure we do not accumulate those over time, as it will affect k8s api server)
2. Ensure we delete volume snapshot object from cloud provider

So this code ensures that. That is why it attempts to talk to all existing clusters and clean it up from all of them. 
If we are no longer connected to some cluster, I think that is fine, as I would assume that cluster is getting decommissioned.

File: components/server/src/workspace/workspace-deletion-service.ts
Comment (sagor999) [reply components/server/src/workspace/workspace-deletion-service.ts]: Fixed

File: components/server/src/workspace/garbage-collector.ts
Comment (sagor999) [reply components/server/src/workspace/garbage-collector.ts:138]: Good point. Added limit. Reusing existing config param for this, assuming it is fine.

File: components/gitpod-db/src/typeorm/workspace-db-impl.ts
Comment (sagor999) [reply components/gitpod-db/src/typeorm/workspace-db-impl.ts:782]: Fixed.

File: components/gitpod-db/src/typeorm/migration/1654628106102-VolumeSnapshotAddWSId.ts
Comment (sagor999) [reply components/gitpod-db/src/typeorm/migration/1654628106102-VolumeSnapshotAddWSId.ts]: Fixed.

**Review by jenting — APPROVED** [review]
LGTM 🛹 🚀

File: components/server/src/workspace/workspace-deletion-service.ts
Comment (geropl) [reply components/server/src/workspace/workspace-deletion-service.ts:142]: > We need to do two things now:

Ah, that explains it. Just to ensure we're on the same page: `ws-manager` does _not_ remove the PVC object from the control plane when the corresponding workspace is stopped?

> Ensure we delete volume snapshot object from cloud provider
I thought we just need to do this, and also wondered if this was something we put into content-service...? :thinking:

Comment (geropl) [reply components/server/src/workspace/workspace-deletion-service.ts:142]: Sorry to "block" the merging of the PR here, but I feel I'm still missing some context to understand why the PVC lifecycle looks the way it does.

Comment (jenting) [reply components/server/src/workspace/workspace-deletion-service.ts:142]: > Just to ensure we're on the same page: `ws-manager` does _not_ remove the PVC object from the control plane when the corresponding workspace is stopped?

No, when the workspace stopped, the PVC object would be removed after the VolumeSnapshot object is ready to use. (Which means the PVC snapshot is done).

> > Ensure we delete volume snapshot object from cloud provider
> 
> I thought we just need to do this, and also wondered if this was something we put into content-service...? 🤔

The lifecycle of PersistentVolumeClaim/VolumeSnapshot is managed by ws-manager.

Comment (sagor999) [reply components/server/src/workspace/workspace-deletion-service.ts:142]: @geropl 
>Just to ensure we're on the same page: ws-manager does not remove the PVC object from the control plane when the corresponding workspace is stopped?

it does. There are two objects in play here. PVC is an actual storage object, it is only alive while workspace is running. When workspace is stopped, wsman will "convert" PVC into VolumeSnapshot object that is stored in cloud provider (outside of k8s control).

When we open workspace, PVC is created and restored from VolumeSnapshot object.

When it is time to delete\GC volume snapshot, we might have same VolumeSnapshot across multiple clusters (prebuild that was opened in different clusters for example). So we want to clean up that volume snapshot object from all clusters (as otherwise we will be accumulating a lot of them over the lifecycle of the cluster), and then also remove actual snapshot from cloud provider as well.

All this work is done by wsman currently.

It will be better to move that into content service indeed, but currently it is not possible, because content service still lives in webapp cluster, and should be migrated into workspace clusters first.

**Comment by jenting** [comment]
Hey Pavel, I re-run this PR to test; I found that after the workspace stopped, I clicked `Delete Workspace` on the GUI. However, the VolumeSnapshot is still there. So I suspect the VolumeSnapshot should be deleted by the ws-manager as well. Correct? Also, I see the loom video doesn't demo this part.

**Review by aledbf — APPROVED** [review]

**Comment by sagor999** [comment]
@jenting it will be deleted when GC runs and deletes workspace for real. 
I simulated that part manually by updating workspace delete timestamp to be one week old, then GC ran and cleaned everything up. 
But thank you for double checking! ❤️

File: components/server/src/workspace/workspace-deletion-service.ts
Comment (geropl) [reply components/server/src/workspace/workspace-deletion-service.ts:142]: Thx @sagor999 for syncing! :+1: 
I understood that:
 - workspace takes care of deleting the actual [disk snapshots](https://cloud.google.com/compute/docs/disks/snapshots?hl=en)
 - we keep the cluster-local VolumeSnapshots and VolumeSnapshotContent in the cluster to optimize for fast workspace starts
 - but to avoid cluttering the clusters too much, we aim to remove those here whenever their disk snapshot get deleted

It would be nice if we added some comments along those lines to the GC. :pray:

File: components/server/src/workspace/workspace-deletion-service.ts
Comment (geropl) [inline components/server/src/workspace/workspace-deletion-service.ts:148]: nit: Already mentioned it, but just to highlight: Especially this part feels a tad brittle, and it would be nice to find a better way to handle/encapsulate that logic.

**Review by geropl — APPROVED** [review]
LGTM :+1:

**Pull request merged**
